### DISCUSSION TITLE
Running from another go routine (systray)? glfw: DestroyWindow failed: Access is denied.

### DISCUSSION THREAD

**Discussion by phrozen** [hajimehoshi/ebiten#3156] [Q&A] [answered]
I am doing something similar to the mascot example in the repo, but I want to be able to control the ebiten app window (create and destroy cycles) with a systray icon app. I can create everything in the goroutine just fine, but when closing the window (by returning an `error `on the `Update()` call) I get a log with `glfw: DestroyWindow failed: Access is denied.`. This prevents me from creating everything again to "re launch" the ebiten app from scratch, program immediately panics with: `panic: ui: u.window must not exist at createWindow`.

I tried using `runtime.LockOSThread()` to no avail. Is there any workaround for this? Is it possible to create and destroy the whole app from within a goroutine?

**Comment by hajimehoshi** [comment, marked as answer]
There is a restriction that `RunGame` must be called from the main goroutine: https://pkg.go.dev/github.com/hajimehoshi/ebiten/v2#RunGame  Were you trying to call this from different goroutines?

**Reply by phrozen** [reply]
I guess I cannot get around that restriction and still manage to create/destroy the app completely from the systray app as both apps (systray/ebiten) need to run on the main thread.

```go
// This runs inside the 'onReady' func passed to systray.Run() which is called inside the main func
// Show clock handler
go func() {
    for {
	    <-show.ClickedCh
	    show.Disable()
	    fmt.Println("Showing clock")
	    cfg := clock.NewDefaultConfig()
	    clk := clock.NewClock(cfg)
	    ebiten.SetWindowSize(clk.Width(), clk.Height())
	    ebiten.SetWindowTitle("Digital Clock")
	    ebiten.SetWindowResizingMode(ebiten.WindowResizingModeDisabled)
	    ebiten.SetWindowFloating(true)
	    ebiten.SetWindowDecorated(false)
	    ebiten.SetTPS(30)
	    op := &ebiten.RunGameOptions{}
	    op.ScreenTransparent = true
	    op.SkipTaskbar = true
	    if err := ebiten.RunGameWithOptions(clk, op); err != nil {
		    log.Println("error:", err)
	    }
	    show.Enable()
    }
}()
```

Alternatively, I can run the app in the `main` thread (by calling it on the may onReady func, and play around with `ebiten.MinimizeWindow()` and `ebiten.RestoreWindow()` functions. Which works... but my main issue with my app, is that even if it is a very basic clock overlay (no decorations, transparent, no title, etc...), and shows only the time, CPU usage for that very basic example is around ~8-10% on Windows AMD Ryzen 5800X, which is quite high for that purpose. So, I wanted to destroy everything when not showing the window. If you have any idea on how to lower CPU usage, I might as well leave it running the whole time, but as it is, is too heavy.

**Reply by phrozen** [reply]
As a side note on CPU usage, minimizing does nothing to it, and lowering the TPS to things like 5 still has the same CPU usage (while affecting the dragging animation). Tracking the time in `Draw()` to only redraw every 500ms does nothing to CPU usage as well. I have tried optimizing everything I can think off, but is still around 8% consistently with some spikes up to 12%. The app only loads a font, calculates the size of the window, and just draws the text, nothing more, which is why I believe usage is quite high for that basic example.

**Reply by hajimehoshi** [reply]
Probably I'd use multiple processes as one Ebitengine process can treat one window.

For CPU usages, you might be able to mitigate it by using `ebiten.SetScreenClearedEveryFrame(false)` and update the screen only when necesasry. Even with this, Ebitengine applications tend to use relatively high CPU usages.

**Reply by phrozen** [reply]
Thanks! I will give that one a try and check if it helps. I'll have everything up in the a repo soon. I can ask again later with concrete code examples. Is there a Discord or some other community to reach you? Or is this the preferred method?

**Reply by hajimehoshi** [reply]
>  Is there a Discord or some other community to reach you? Or is this the preferred method?

Here you are :) https://discord.gg/3tVdM5H8cC

**Reply by phrozen** [reply]
Lovely! That last thing (using `ebiten.SetScreenClearedEveryFrame(false)`) and keeping track of last rendered time to only update every 500ms+ brought it down to around 2%, which I guess is way more acceptable (75% CPU usage reduction), will keep optimizing and see where else I can gain a few things. Closing this now.
